File: test/technology/qca/test_qca_layout.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/technology/qca/layout.hpp`.
 * @author Marcel Walter (marcelwa)
 */

#include <catch2/catch_test_macros.hpp>

#include <fiction/layouts/clocking_scheme.hpp>
#include <fiction/layouts/coordinates.hpp>
#include <fiction/technology/qca/layout.hpp>

#include <stdexcept>

using namespace fiction;
using namespace fiction::layouts;
using namespace fiction::qca;

TEST_CASE("QCA cell types", "[qca-layout]")
{
    CHECK(is_constant(cell_type::CONST_0));
    CHECK(is_constant(cell_type::CONST_1));
    CHECK(!is_constant(cell_type::NORMAL));
    CHECK(!is_constant(cell_type::INPUT));
}

TEST_CASE("QCA layout construction", "[qca-layout]")
{
    CHECK_THROWS_AS((layout{{2, 2}, "", 0, 1}), std::invalid_argument);
    CHECK_THROWS_AS((layout{{2, 2}, clocking::twoddwave<layout>(), "", 1, 0}), std::invalid_argument);

    const layout lyt{{4, 4, 1}, clocking::use<layout>(), "crossing", 5, 5};

    CHECK(lyt.x() == 4);
    CHECK(lyt.z() == 1);
    CHECK(lyt.get_layout_name() == "crossing");
    CHECK(lyt.is_clocking_scheme(clocking::USE_NAME));
    CHECK(lyt.get_tile_size_x() == 5);
    CHECK(lyt.get_tile_size_y() == 5);
}

TEST_CASE("QCA cell modes on the crossing layer", "[qca-layout]")
{
    layout lyt{{4, 4, 1}, "crossover"};

    lyt.assign_cell_type({2, 2, 0}, cell_type::NORMAL);
    lyt.assign_cell_type({2, 1, 1}, cell_type::NORMAL);
    lyt.assign_cell_type({2, 2, 1}, cell_type::NORMAL);

    lyt.assign_cell_mode({2, 1, 1}, cell_mode::VERTICAL);
    lyt.assign_cell_mode({2, 2, 1}, cell_mode::CROSSOVER);

    CHECK(lyt.is_crossing_layer({2, 2, 1}));
    CHECK(lyt.num_cells() == 3);
    CHECK(lyt.get_cell_mode({2, 2, 0}) == cell_mode::NORMAL);
    CHECK(lyt.get_cell_mode({2, 1, 1}) == cell_mode::VERTICAL);
    CHECK(lyt.get_cell_mode({2, 2, 1}) == cell_mode::CROSSOVER);

    SECTION("normal mode removes a stored mode")
    {
        lyt.assign_cell_mode({2, 1, 1}, cell_mode::NORMAL);
        CHECK(lyt.get_cell_mode({2, 1, 1}) == cell_mode::NORMAL);
    }
    SECTION("removing a cell removes its mode")
    {
        lyt.assign_cell_type({2, 2, 1}, cell_type::EMPTY);
        lyt.assign_cell_type({2, 2, 1}, cell_type::NORMAL);
        CHECK(lyt.get_cell_mode({2, 2, 1}) == cell_mode::NORMAL);
    }
}

TEST_CASE("QCA synchronization elements belong to clock zones", "[qca-layout]")
{
    layout original{{5, 5, 1}, clocking::twoddwave<layout>(), "cells", 2, 2};

    original.assign_synchronization_element({1, 1}, 2);

    CHECK(original.num_se() == 1);
    CHECK(original.get_synchronization_element({2, 2}) == 2);
    CHECK(original.get_synchronization_element({3, 3}) == 2);
    CHECK(original.is_synchronization_element({3, 2, 1}));
    CHECK(original.get_synchronization_element({4, 4}) == 0);
    CHECK(!original.is_synchronization_element({1, 1}));

    SECTION("copies are independent")
    {
        auto copy = original;
        CHECK(copy == original);

        copy.assign_synchronization_element({1, 1}, 0);
        copy.assign_clock_number({1, 1}, 0);
        copy.assign_cell_type({0, 0}, cell_type::NORMAL);
        copy.assign_cell_mode({0, 0}, cell_mode::ROTATED);

        CHECK(copy.num_se() == 0);
        CHECK(original.get_synchronization_element({3, 3}) == 2);
        CHECK(original.get_clock_number({3, 3}) == 2);
        CHECK(original.is_empty());
        CHECK(copy != original);
    }
}

File: test/python/qca/test_qca_layout.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

from __future__ import annotations

import copy

import pytest

from mnt.pyfiction.layouts.coords import offset_coordinate
from mnt.pyfiction.qca import qca_cell_mode, qca_cell_type, qca_layout


def test_cell_types_and_modes() -> None:
    assert str(qca_cell_type.CONST_0) == "qca_cell_type.CONST_0"
    assert str(qca_cell_mode.CROSSOVER) == "qca_cell_mode.CROSSOVER"


def test_geometry_is_cartesian() -> None:
    layout = qca_layout(offset_coordinate(9, 9, 1))

    for t in layout.coordinates():
        assert t <= offset_coordinate(9, 9, 1)
        assert layout.is_within_bounds(t)

    for t in layout.ground_coordinates():
        assert t.z == 0

    for t in layout.adjacent_coordinates(offset_coordinate(2, 2)):
        assert t in [(1, 2), (2, 1), (3, 2), (2, 3)]


def test_cell_type_and_mode_assignment() -> None:
    layout = qca_layout(offset_coordinate(4, 4, 1), "OPEN", "crossing")

    assert layout.is_empty()
    assert layout.get_layout_name() == "crossing"

    layout.assign_cell_type(offset_coordinate(0, 2), qca_cell_type.INPUT)
    layout.assign_cell_type(offset_coordinate(2, 2), qca_cell_type.NORMAL)
    layout.assign_cell_type(offset_coordinate(2, 2, 1), qca_cell_type.NORMAL)
    layout.assign_cell_type(offset_coordinate(4, 2), qca_cell_type.OUTPUT)
    layout.assign_cell_mode(offset_coordinate(2, 2, 1), qca_cell_mode.CROSSOVER)
    layout.assign_cell_name(offset_coordinate(0, 2), "a")

    assert layout.num_cells() == 4
    assert layout.num_pis() == 1
    assert layout.num_pos() == 1
    assert layout.is_pi(offset_coordinate(0, 2))
    assert layout.pis() == [(0, 2)]
    assert layout.get_cell_name(offset_coordinate(0, 2)) == "a"
    assert layout.get_cell_type(offset_coordinate(2, 2, 1)) == qca_cell_type.NORMAL
    assert layout.get_cell_mode(offset_coordinate(2, 2, 1)) == qca_cell_mode.CROSSOVER
    assert layout.get_cell_mode(offset_coordinate(2, 2)) == qca_cell_mode.NORMAL
    assert layout.is_empty_cell(offset_coordinate(1, 1))

    layout.assign_cell_type(offset_coordinate(2, 2, 1), qca_cell_type.EMPTY)
    assert layout.is_empty_cell(offset_coordinate(2, 2, 1))
    assert layout.get_cell_mode(offset_coordinate(2, 2, 1)) == qca_cell_mode.NORMAL


def test_clock_zones_and_synchronization_elements() -> None:
    """Clock numbers and synchronization elements belong to clock zones, i.e., to tiles of cells."""
    layout = qca_layout(offset_coordinate(4, 4), "2DDWave", "", 2, 2)
    assert layout.get_tile_size_x() == 2
    assert layout.is_clocking_scheme("2DDWAVE")
    with pytest.raises(ValueError, match="positive"):
        layout.set_tile_size_x(0)
    with pytest.raises(ValueError, match="positive"):
        qca_layout(offset_coordinate(4, 4), "2DDWave", "", 0, 1)

    assert layout.get_clock_zone(offset_coordinate(3, 2)) == (1, 1)
    layout.assign_clock_number(offset_coordinate(1, 1), 3)
    layout.assign_synchronization_element(offset_coordinate(1, 1), 2)
    assert layout.get_clock_number(offset_coordinate(2, 2)) == 3
    assert layout.get_synchronization_element(offset_coordinate(3, 3)) == 2
    assert layout.get_synchronization_element(offset_coordinate(1, 1)) == 0
    assert layout.num_se() == 1

    duplicate = copy.copy(layout)
    assert duplicate == layout
    duplicate.assign_clock_number(offset_coordinate(1, 1), 0)
    duplicate.assign_synchronization_element(offset_coordinate(1, 1), 0)
    assert layout.get_clock_number(offset_coordinate(2, 2)) == 3
    assert layout.num_se() == 1

    layout.replace_clocking_scheme("USE")
    assert layout.is_clocking_scheme("USE")
    assert layout.get_clocking_scheme_name() == "USE"
    with pytest.raises(ValueError, match="Unknown clocking scheme"):
        layout.replace_clocking_scheme("3DDWave")
    with pytest.raises(ValueError, match="clocking scheme"):
        qca_layout(offset_coordinate(4, 4), "3DDWave")
